presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,374 @@
+                if (lowerBoundComparison < 0) {
+                    // smaller than minimum representable value
+                    switch (operator) {
+                        case NOT_EQUAL:
+                        case GREATER_THAN:
+                        case GREATER_THAN_OR_EQUAL:
+                            return trueIfNotNull(cast.getExpression());
+                        case EQUAL:
+                        case LESS_THAN:
+                        case LESS_THAN_OR_EQUAL:
+                            return falseIfNotNull(cast.getExpression());
+                        case IS_DISTINCT_FROM:
+                            return TRUE_LITERAL;
+                        default:
+                            throw new UnsupportedOperationException("not yet implemented: " + operator);
+                    }
+                }
+
+                if (lowerBoundComparison == 0) {
+                    // equal to min representable value
+                    switch (operator) {
+                        case LESS_THAN:
+                            return falseIfNotNull(cast.getExpression());
+                        case LESS_THAN_OR_EQUAL:
+                            return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        case GREATER_THAN_OR_EQUAL:
+                            return trueIfNotNull(cast.getExpression());
+                        case GREATER_THAN:
+                            return new ComparisonExpression(NOT_EQUAL, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        case EQUAL:
+                        case NOT_EQUAL:
+                        case IS_DISTINCT_FROM:
+                            return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                        default:
+                            throw new UnsupportedOperationException("not yet implemented: " + operator);
+                    }
+                }
+            }
+
+            Signature targetToSource;
+            try {
+                targetToSource = metadata.getFunctionRegistry().getCoercion(literalType, sourceType);
+            }
+            catch (OperatorNotFoundException e) {
+                // Without a cast between target -> source, there's nothing we can do
+                return expression;
+            }
+
+            Object literalInSourceType;
+            try {
+                literalInSourceType = functionInvoker.invoke(
+                        targetToSource,
+                        session.toConnectorSession(),
+                        literalValue);
+            }
+            catch (PrestoException e) {
+                // can't convert the literal to the source type. This means the literal
+                // is not representable in the the source type, so we can infer no value in
+                // source is equal to the literal
```

Comment:
This sounds reasonable. Can we have a kill switch though?

Reply:
I was actually going to add a kill switch for the whole feature.

Reply:
Sure, a switch for the whole feature is preferred. Sorry, i wasn't clear.

Reply:
However, I agree that there's something odd about this case. It is theoretically possible for an implicit cast to exist between a type S and T without a cast being defined for T->S. Also, it's possible (though arguably broken) for a `cast(s as T)` to succeed, a cast T->S to exist, but `cast(cast(s as T) as S)` to fail due to a bad implementation. In that case, the transformation from `cast(s as T) = t` to `if(s is null, null, false)` would be incorrect.

I'll change this case to be more conservative.

Reply:
Hm... I am not sure about how these transformations preserve _failures_. For example, `cast(nan() as integer)` is a failure, so in theory for `d :: double`, you can unwrap `cast(d as integer) = 42`. 
I thought about that earlier. However, I think we already do optimizations that do not preserve failures (eg division by zero vs prune unreferenced outputs), so I think this is OK to optimize a query like that.

Reply:
Double->integer is not an implicit cast, so the rule wouldn’t apply in that case.

I have a bunch of ideas on how to do unwrap in that case, but it’s a very different set of rules, and we’ll probably need functions for computing the next/previous of a given value and a way to synthesize failures

Reply:
The premise for the transformation in this PR is that implicit casts *can’t* fail (they are injective), so there’s no need to reason about failures when casting from the “narrower” to the “wider” type